File: packages/anki/sync/state.py
"""SQLite WAL state database for tracking sync state."""

from __future__ import annotations

import sqlite3
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

import structlog

if TYPE_CHECKING:
    from pathlib import Path
    from types import TracebackType

log = structlog.get_logger()


@dataclass(frozen=True, slots=True)
class CardState:
    """Tracked state for a single card."""

    slug: str
    content_hash: str
    anki_guid: int | None = None
    note_type: str = ""
    source_path: str = ""
    synced_at: float = 0.0


class StateDB:
    """SQLite WAL database for tracking sync state."""

    def __init__(self, db_path: Path) -> None:
        self._db_path = db_path
        self._conn = sqlite3.connect(str(db_path))
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA foreign_keys=ON")
        self._create_tables()
        log.debug("state_db.opened", path=str(db_path))

    def _create_tables(self) -> None:
        self._conn.execute(
            """
            CREATE TABLE IF NOT EXISTS card_state (
                slug TEXT PRIMARY KEY,
                content_hash TEXT NOT NULL,
                anki_guid INTEGER,
                note_type TEXT NOT NULL DEFAULT '',
                source_path TEXT NOT NULL DEFAULT '',
                synced_at REAL NOT NULL DEFAULT 0.0
            )
            """
        )
        self._conn.commit()

    def get(self, slug: str) -> CardState | None:
        """Get card state by slug."""
        row = self._conn.execute(
            "SELECT slug, content_hash, anki_guid, note_type, source_path, synced_at "
            "FROM card_state WHERE slug = ?",
            (slug,),
        ).fetchone()
        if row is None:
            return None
        return CardState(
            slug=row[0],
            content_hash=row[1],
            anki_guid=row[2],
            note_type=row[3],
            source_path=row[4],
            synced_at=row[5],
        )

    def get_all(self) -> tuple[CardState, ...]:
        """Get all card states."""
        rows = self._conn.execute(
            "SELECT slug, content_hash, anki_guid, note_type, source_path, synced_at "
            "FROM card_state ORDER BY slug"
        ).fetchall()
        return tuple(
            CardState(
                slug=r[0], content_hash=r[1], anki_guid=r[2],
                note_type=r[3], source_path=r[4], synced_at=r[5],
            )
            for r in rows
        )

    def upsert(self, state: CardState) -> None:
        """Insert or update card state."""
        self._conn.execute(
            """
            INSERT INTO card_state (slug, content_hash, anki_guid, note_type, source_path, synced_at)
            VALUES (?, ?, ?, ?, ?, ?)
            ON CONFLICT (slug) DO UPDATE SET
                content_hash = excluded.content_hash,
                anki_guid = excluded.anki_guid,
                note_type = excluded.note_type,
                source_path = excluded.source_path,
                synced_at = excluded.synced_at
            """,
            (
                state.slug,
                state.content_hash,
                state.anki_guid,
                state.note_type,
                state.source_path,
                state.synced_at,
            ),
        )
        self._conn.commit()

    def delete(self, slug: str) -> None:
        """Delete card state by slug."""
        self._conn.execute("DELETE FROM card_state WHERE slug = ?", (slug,))
        self._conn.commit()

    def get_by_source(self, source_path: str) -> tuple[CardState, ...]:
        """Get all card states for a source path."""
        rows = self._conn.execute(
            "SELECT slug, content_hash, anki_guid, note_type, source_path, synced_at "
            "FROM card_state WHERE source_path = ? ORDER BY slug",
            (source_path,),
        ).fetchall()
        return tuple(
            CardState(
                slug=r[0], content_hash=r[1], anki_guid=r[2],
                note_type=r[3], source_path=r[4], synced_at=r[5],
            )
            for r in rows
        )

    def close(self) -> None:
        """Close database connection."""
        self._conn.close()
        log.debug("state_db.closed", path=str(self._db_path))

    def __enter__(self) -> StateDB:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        self.close()


def _now() -> float:
    """Current time as float (for synced_at)."""
    return time.time()

File: crates/anki-sync/src/state.rs
use std::path::Path;
use std::sync::Mutex;

use rusqlite::{Connection, Row};

/// Errors from StateDB operations.
#[derive(Debug, thiserror::Error)]
pub enum StateDbError {
    #[error("sqlite error: {0}")]
    Sqlite(#[from] rusqlite::Error),

    #[error("mutex poisoned")]
    MutexPoisoned,
}

/// Tracked state for a single card.
#[derive(Debug, Clone, PartialEq)]
pub struct CardState {
    pub slug: String,
    pub content_hash: String,
    pub anki_guid: Option<i64>,
    pub note_type: String,
    pub source_path: String,
    pub synced_at: f64,
}

impl CardState {
    /// Map a SQLite row to a `CardState`.
    ///
    /// Expects columns in order: slug, content_hash, anki_guid, note_type, source_path, synced_at.
    fn from_row(row: &Row<'_>) -> rusqlite::Result<Self> {
        Ok(Self {
            slug: row.get(0)?,
            content_hash: row.get(1)?,
            anki_guid: row.get(2)?,
            note_type: row.get(3)?,
            source_path: row.get(4)?,
            synced_at: row.get(5)?,
        })
    }
}

/// SQLite WAL database for tracking per-card sync state.
pub struct StateDB {
    conn: Mutex<Connection>,
}

impl StateDB {
    /// Open or create the state database at `db_path`.
    /// Enables WAL mode and foreign keys. Creates the `card_state` table if needed.
    pub fn open(db_path: impl AsRef<Path>) -> Result<Self, StateDbError> {
        let conn = Connection::open(db_path.as_ref())?;

        conn.pragma_update(None, "journal_mode", "WAL")?;
        conn.pragma_update(None, "foreign_keys", "ON")?;

        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS card_state (
                slug TEXT PRIMARY KEY,
                content_hash TEXT NOT NULL,
                anki_guid INTEGER,
                note_type TEXT NOT NULL DEFAULT '',
                source_path TEXT NOT NULL DEFAULT '',
                synced_at REAL NOT NULL DEFAULT 0.0
            );",
        )?;

        Ok(Self {
            conn: Mutex::new(conn),
        })
    }

    /// Get card state by slug, or `None` if not found.
    pub fn get(&self, slug: &str) -> Result<Option<CardState>, StateDbError> {
        let conn = self.conn.lock().map_err(|_| StateDbError::MutexPoisoned)?;
        match conn.query_row(
            "SELECT slug, content_hash, anki_guid, note_type, source_path, synced_at
                 FROM card_state WHERE slug = ?1",
            [slug],
            CardState::from_row,
        ) {
            Ok(state) => Ok(Some(state)),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(StateDbError::Sqlite(e)),
        }
    }

    /// Get all card states, sorted by slug.
    pub fn get_all(&self) -> Result<Vec<CardState>, StateDbError> {
        let conn = self.conn.lock().map_err(|_| StateDbError::MutexPoisoned)?;
        let mut stmt = conn.prepare(
            "SELECT slug, content_hash, anki_guid, note_type, source_path, synced_at
                 FROM card_state ORDER BY slug",
        )?;

        let rows = stmt.query_map([], CardState::from_row)?;
        let mut results = Vec::new();
        for row in rows {
            results.push(row?);
        }
        Ok(results)
    }

    /// Insert or update a card state (upsert on slug).
    pub fn upsert(&self, state: &CardState) -> Result<(), StateDbError> {
        let conn = self.conn.lock().map_err(|_| StateDbError::MutexPoisoned)?;
        conn.execute(
            "INSERT INTO card_state (slug, content_hash, anki_guid, note_type, source_path, synced_at)
                 VALUES (?1, ?2, ?3, ?4, ?5, ?6)
                 ON CONFLICT(slug) DO UPDATE SET
                    content_hash = excluded.content_hash,
                    anki_guid = excluded.anki_guid,
                    note_type = excluded.note_type,
                    source_path = excluded.source_path,
                    synced_at = excluded.synced_at",
            rusqlite::params![
                state.slug,
                state.content_hash,
                state.anki_guid,
                state.note_type,
                state.source_path,
                state.synced_at,
            ],
        )?;
        Ok(())
    }

    /// Delete card state by slug.
    pub fn delete(&self, slug: &str) -> Result<(), StateDbError> {
        let conn = self.conn.lock().map_err(|_| StateDbError::MutexPoisoned)?;
        conn.execute("DELETE FROM card_state WHERE slug = ?1", [slug])?;
        Ok(())
    }

    /// Get all card states for a given source path, sorted by slug.
    pub fn get_by_source(&self, source_path: &str) -> Result<Vec<CardState>, StateDbError> {
        let conn = self.conn.lock().map_err(|_| StateDbError::MutexPoisoned)?;
        let mut stmt = conn.prepare(
            "SELECT slug, content_hash, anki_guid, note_type, source_path, synced_at
                 FROM card_state WHERE source_path = ?1 ORDER BY slug",
        )?;

        let rows = stmt.query_map([source_path], CardState::from_row)?;
        let mut results = Vec::new();
        for row in rows {
            results.push(row?);
        }
        Ok(results)
    }

    /// Close the database connection.
    pub fn close(self) {
        // Connection is closed when dropped
    }
}
